### ISSUE TITLE
Different behavior of render depending on the order of the bindings

### ISSUE THREAD

**Issue by DiegoCardoso** [Polymer/lit-html#1648]
# Description
This was first noticed while using `render` with a Polymer 3 component. I couldn't quickly reproduce the same behavior with a LitElement, but it may still occur.

When a `TemplateInstance` has bindings that are somehow connected, it may end up with different results depending on the order on which they are declared.

For instance, say you have a P3 element with a property which uses the `notify` feature. If you declare the property and the event associated with its change, the result will differ if the property declaration comes first or last. That seems to be that, because notify makes the call to event listeners synchronously to the property change, if the property binding comes first, then the listener will be called before it's updated.

# Steps to Reproduce

1. Write this code

```ts
import { render, html } from "lit-html";

import "./my-element.js";

// Define a template function
const myTemplate = name =>
  html`
    <my-element
      .name=${name}
      @name-changed=${e => console.log("event after name", name)}
    ></simple-greeting>

    <my-element
      @name-changed=${e => console.log("event before name", name)}
      .name=${name}
    ></simple-greeting>
  `;

const container = document.getElementById("container");
const buttons = document.getElementById("buttons");

// Render the template with some data
render(myTemplate("initial state"), container);

render(
  html`
    <button @click=${() => render(myTemplate("foo"), container)}>
      Name: foo
    </button>
    <button @click=${() => render(myTemplate("bar"), container)}>
      Name: bar
    </button>
  `,
  buttons
);
```

// my-element.js
```js
import { PolymerElement, html } from "@polymer/polymer";

// Define an element class
class MyElement extends PolymerElement {
  // Define public API properties
  static get properties() {
    return {
      name: {
        type: Boolean,
        notify: true
      }
    };
  }

  // Define the element's template
  static get template() {
    return html`
      <div>Name is {{name}}</div>
    `;
  }
}

// Register the element with the browser
customElements.define("my-element", MyElement);

```

2. Open the console to see the results
![image](https://user-images.githubusercontent.com/262432/109862477-3f32a680-7c3f-11eb-8ddd-53b3f9fd647b.png)

## Live Reproduction Link
https://stackblitz.com/edit/lit-html-frxdw2?file=src%2Findex.js

## Expected Results
Output should be the same regardless of the order on which the bindings are declared
```sh
event after foo
event before foo
```

## Actual Results
The old listener is called before it's updated, leading to unexpected results.
```sh
event after initial state
event before foo
```

## Browsers Affected

- [x] Chrome (only tested on Chrome but it should apply to all browers)
- [ ] Firefox
- [ ] Edge
- [ ] Safari 11
- [ ] Safari 10
- [ ] IE 11

**Comment by justinfagnani** [comment]
The ordering of lit-html binding execution matching the source ordering by design and fundamental to the way that execution is ordered in JavaScript. Template expressions are evaluated in JS in source order, and we can't change that. So we process template parts in their source order too. It's simpler and less code than some kind of re-ordering based on part type, plus we try to match the semantics of what the equivalent code would be when possible.

Note that Polymer's two-way data-binding protocol is the thing that's order sensitive here, because it fires events synchronously, and it would be sensitive in imperative code as well. ie:

```ts
const myEl = document.createElement('my-element');
myEl.name = 'foo';
myEl.addEventListener('name-changed', (e) => console.log(myEl.name));
```

So in the Polymer-interop case it's important that property-change event handlers come first in the element, _if_ you need to synchronize with data outside the element _and_ you're replacing the event handler every render - as you have in your example with a closure that closes over the name parameter to the template function.

If you read the name off the element instead, then the difference in order is not observable:
```ts
// Define a template function
const myTemplate = name =>
  html`
    <my-element
      .name=${name}
      @name-changed=${e => console.log("event after name", e.target.name)}
    ></simple-greeting>

    <my-element
      @name-changed=${e => console.log("event before name", e.target.name)}
      .name=${name}
    ></simple-greeting>
  `;
```

The subtle difference is that in your example the "event after name" listener that fires is actually the listener from the previous render, which closed over the previous `name` parameter.

You can see that if you add a counter:

```ts
// Define a template function
let i = 0;
const myTemplate = name => {
  const j = i++;
  return html`
    <my-element
      .name=${name}
      @name-changed=${e => console.log("event after name", j, name)}
    ></simple-greeting>

    <my-element
      @name-changed=${e => console.log("event before name", j, name)}
      .name=${name}
    ></simple-greeting>
  `;
};
```

But to sum up, we're very sparing with adding more code to core, especially when it would change behavior from what seems apparent in the source. It's pretty explainable that bindings execute in source order, and harder to explain otherwise.

**Comment by DiegoCardoso** [comment]
Thanks for the answer.

I was aware that the previous listener was being called instead of the new one which holds the previous value of `name`. I wanted to check whether that was expected or not, but it's clear to me now.

Later, I tried to use the original component where I found this "issue" by creating and assigning the properties/listeners imperatively and I came across the same issue if I changed the property before assigning the new listener (the same case you provided as an example). On my case, the event `detail` doesn't hold the value I needed, so that's why I needed to access the object directly from the listener.
